### PR TITLE
Integrate keytool for keystores

### PR DESCRIPTION
**Pull request by snarlysodboxer**
Update certificate generation to support multiple algorithms.
Update certificate generation to support the example config (it already supported the certs needed for the mutating webhook.)
Add testing instructions and a Dockerfile.testing file for testing.
Update config validation to use Kubernetes object validation wherever possible, only using the v10 validation library for custom validation.
Update example configs, and corresponding types and code.
Update test rigs according to the new example config.
Update cert generation in a way that can be done systematically using values previously stored in the memory store; this enables the recursive generation and regeneration used by this application.
Add tests that output the test rigs so they can be seen in YAML for understanding.
Add a test proving that when not using an external secret manager, we won't regenerate a keystore if it's value already exists and it's alias values are missing.
Add keytool import functions.
Add tests for each of the keytool import functions.
Other testing improvements.
Support regeneration of SSH public keys, should they be lost.
Update and simplify dependency tree calculation, including supporting new config.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -91,39 +91,39 @@ func main() {
...
 
 	setupLog.Info("patching webhook secret", "name", secretName)
-	if err := controllers.PatchWebhookSecret(rootCA.CAPem, cert, key, secretName, namespace); err != nil {
+	if err := controllers.PatchWebhookSecret(rootCA.CertPEM, leafCert.CertPEM, leafCert.PrivateKeyPEM, secretName, namespace); err != nil {
 		setupLog.Error(err, "Unable to patch secret")
 	}
 
 	setupLog.Info("patching validating webhook", "name", validatingWebhookConfigurationName)
-	if err := controllers.PatchValidatingWebhookConfiguration(rootCA.CAPem, validatingWebhookConfigurationName); err != nil {
+	if err := controllers.PatchValidatingWebhookConfiguration(rootCA.CertPEM, validatingWebhookConfigurationName); err != nil {
```
Comment (jrcast) [inline main.go:114]: good catch

File: config/crd/bases/secret-agent.secrets.forgerock.io_secretagentconfigurations.yaml [hunk 2]
```diff
@@ -148,14 +147,16 @@ spec:
                           type: array
                         type:
                           description: KeyConfigType Specifies which key type to use
-                            Only one of the following types may be specified.
```
[the remark this answers is not in the document]
Comment (snarlysodboxer) [reply config/crd/bases/secret-agent.secrets.forgerock.io_secretagentconfigurations.yaml:151]: it's still true, just redundant with the word "enum", is it not?

**Comment by jrcast** [comment]
Not sure why the docker build failed. Did you replace a commit while it was running? 🤷

File: api/v1alpha1/secretagentconfiguration_webhook.go [hunk 3]
```diff
@@ -92,90 +92,110 @@ func ConfigurationStructLevelValidator(sl validator.StructLevel) {
 		if config.AppConfig.GCPProjectID == "" {
 			sl.ReportError(config.AppConfig.GCPProjectID, "gcpProjectID", "GCPProjectID", "emptyGCPProjectID", "")
 		}
-	case SecretsManagerAzure:
```
[the remark this answers is not in the document]
Comment (snarlysodboxer) [reply api/v1alpha1/secretagentconfiguration_webhook.go:95]: I'm not sure how this happened. I pulled in the latest from master before submitting this.

File: controllers/webhook_cabundler.go [hunk 4]
```diff
@@ -26,23 +26,22 @@ func getClientSet(kubeconfig string) (*k8s, error) {
...
 	}
+
 	return
 }
 
-// PatchWebhookSecret patches the named tls secret with the TLS information
-func PatchWebhookSecret(rootCAPem, cert, key []byte, name string, namespace string) (err error) {
-
+// PatchWebhookSecret patches the named TLS secret with the TLS information
+func PatchWebhookSecret(rootCAPem, certPEM, keyPEM []byte, name string, namespace string) (err error) {
```
[the remark this answers is not in the document]
Comment (snarlysodboxer) [reply controllers/webhook_cabundler.go:44]: name and namespace are both of type string, so this is valid, and follows the same shorthand pattern we're using for rootCAPem, certPEM, and keyPEM.

File: api/v1alpha1/secretagentconfiguration_types.go [hunk 5]
```diff
@@ -98,17 +102,20 @@ const (
 	TypePassword     KeyConfigType = "password"
 	TypePrivateKey   KeyConfigType = "privateKey"
 	TypePublicKeySSH KeyConfigType = "publicKeySSH"
-	TypeJCEKS        KeyConfigType = "jceks"
+	TypeCA           KeyConfigType = "ca"
```
[the remark this answers is not in the document]
Comment (snarlysodboxer) [reply api/v1alpha1/secretagentconfiguration_types.go:105]: Yes, it does fix that. It was a by product of the way we changed the example config.

File: api/v1alpha1/secretagentconfiguration_types.go [hunk 6]
```diff
@@ -27,9 +27,11 @@ type SecretAgentConfigurationSpec struct {
 	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
 
-	AppConfig AppConfig `json:"appConfig" validate:"required,dive,required"`
+	// +kubebuilder:validation:Required
+	AppConfig AppConfig `json:"appConfig" yaml:"appConfig,omitempty"`
```
[the remark this answers is not in the document]
Comment (snarlysodboxer) [reply api/v1alpha1/secretagentconfiguration_types.go:31]: Yes, that's right.

File: README.md [hunk 7]
```diff
@@ -18,3 +18,11 @@ To deploy the secret-agent, run:
 ```bash
 kustomize build config/default | kubectl apply -f -
 ```
+
+## Running the tests
+* Some of the tests exercise parts of the code that os.Exec dskeymgr and keytool
```
[the remark this answers is not in the document]
Comment (snarlysodboxer) [reply README.md:23]: Good call. I'll fix.

File: Dockerfile [hunk 8]
```diff
@@ -46,3 +45,8 @@ RUN mkdir -p /opt/gen/secrets/generic/truststore && \
     chown -R forgerock:root /opt/gen
 
 USER forgerock
+
+# Set the entrypoint to /manager with automation,
+#   leave it as bash here to prevent the need for manual override when debugging
```
[the remark this answers is not in the document]
Comment (snarlysodboxer) [reply Dockerfile:50]: Good call

File: .gitignore [hunk 9]
```diff
@@ -24,3 +24,7 @@ bin
 *~
 /.project
 /.vscode/
+
+testConfiguration1.yaml
```
[the remark this answers is not in the document]
Comment (snarlysodboxer) [reply .gitignore:28]: It just prevents the output of the TestOutputTestRigs test from accidentally being committed.

File: Dockerfile (hunk 8 above)
[the remark this answers is not in the document]
Comment (jrcast) [reply Dockerfile:50]: I am curious to know how much smaller the image will be

File: api/v1alpha1/secretagentconfiguration_types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jrcast) [reply api/v1alpha1/secretagentconfiguration_types.go:105]: Ok. 
This PR fixes #18

File: controllers/webhook_cabundler.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jrcast) [reply controllers/webhook_cabundler.go:44]: correct

File: Dockerfile (hunk 8 above)
[the remark this answers is not in the document]
Comment (jrcast) [reply Dockerfile:50]: Linked this PR to #10

File: controllers/webhook_cabundler.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (snarlysodboxer) [reply controllers/webhook_cabundler.go:44]: Oh, I misread the comment at first. I'll fix.

[the remark this answers is not in the document]
Comment (snarlysodboxer) [reply controllers/webhook_cabundler.go]: fixed

File: README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (snarlysodboxer) [reply README.md]: fixed

File: api/v1alpha1/secretagentconfiguration_webhook.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (snarlysodboxer) [reply api/v1alpha1/secretagentconfiguration_webhook.go]: fixed

**Comment by snarlysodboxer** [comment]
> Not sure why the docker build failed. Did you replace a commit while it was running? 🤷

I don't think so. I don't know why this happened, I can't see the build output.  It seems to be gone now that I added a commit and squashed it.

**Comment by jrcast** [comment]
The docker build seems to be broken:
```
Cloning into '.'...
Switched to a new branch 'integrate-keytool'
KernelVersion: 4.4.0-1060-aws
Components: [{u'Version': u'19.03.8', u'Name': u'Engine', u'Details': {u'KernelVersion': u'4.4.0-1060-aws', u'Os': u'linux', u'BuildTime': u'2020-03-11T01:24:30.000000000+00:00', u'ApiVersion': u'1.40', u'MinAPIVersion': u'1.12', u'GitCommit': u'afacb8b7f0', u'Arch': u'amd64', u'Experimental': u'false', u'GoVersion': u'go1.12.17'}}, {u'Version': u'1.2.13', u'Name': u'containerd', u'Details': {u'GitCommit': u'7ad184331fa3e55e52b890ea95e65ba581ae3429'}}, {u'Version': u'1.0.0-rc10', u'Name': u'runc', u'Details': {u'GitCommit': u'dc9208a3303feef5b3839f4323d9beb36df0a9dd'}}, {u'Version': u'0.18.0', u'Name': u'docker-init', u'Details': {u'GitCommit': u'fec3683'}}]
Arch: amd64
BuildTime: 2020-03-11T01:24:30.000000000+00:00
ApiVersion: 1.40
Platform: {u'Name': u'Docker Engine - Community'}
Version: 19.03.8
MinAPIVersion: 1.12
GitCommit: afacb8b7f0
Os: linux
GoVersion: go1.12.17
Starting build of index.docker.io/forgerock/secret-agent:this...
Step 1/13 : FROM golang:1.14-alpine as builder
---> 3289bf11c284
Step 2/13 : WORKDIR /workspace
---> Running in a41a9095aadd
Removing intermediate container a41a9095aadd
---> aab9364095f5
Step 3/13 : COPY go.mod go.sum ./
---> ad4ee1b2a4c6
Step 4/13 : RUN go mod download
---> Running in 07a3370a52bd
Removing intermediate container 07a3370a52bd
---> b6c0824d69db
Step 5/13 : COPY . .
---> 5e3538c37b7d
Step 6/13 : RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -ldflags "-s -w" -a -o manager main.go
---> Running in 374f6b46984a
[91m# command-line-arguments
./main.go:129:64: undefined: cert
./main.go:132:64: undefined: key
[0m
Removing intermediate container 374f6b46984a
The command '/bin/sh -c CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -ldflags "-s -w" -a -o manager main.go' returned a non-zero code: 2
```

Can you try to run:
`make docker-build IMG=forgerock/secret-agent:master`

**Comment by snarlysodboxer** [comment]
> The docker build seems to be broken:
> Step 6/13 : RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -ldflags "-s -w" -a -o manager main.go
> ---> Running in 374f6b46984a
> �[91m# command-line-arguments
> ./main.go:129:64: undefined: cert
> ./main.go:132:64: undefined: key
> �[0m
> Removing intermediate container 374f6b46984a
> The command '/bin/sh -c CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -ldflags "-s -w" -a -o manager main.go' returned a non-zero code: 2
> ```
> 
> Can you try to run:
> `make docker-build IMG=forgerock/secret-agent:master` to test locally

Doh. Fixed.

FWIW, I'm pro getting rid of the `make docker-build` command.
1. It's more typing than the native docker command.
2. If you do know docker, you already know to run `docker build -t gcr.io/forgerock-io/secret-agent:latest .`
3. If you don't know enough about docker, learning our make command won't help you learn one of the most basic things about using docker.
4. We don't have to support that part of the makefile.
5. Giving people two ways to do the same thing can increase confusion or just take extra energy wondering if they're different somehow.

**Comment by jrcast** [comment]
I'd like to keep it. Note that there are other things that happen as part of `make docker-build` including testing, ` go fmt` and `go vet`, generating manifests, etc. 

I would argue one is more likely to mess up (forgetting to generate the CRDs or something along those lines) by not using automation. Yes you can run `make test` + `docker build .... ` to achieve the same result. However, that ends up being more steps and you're still using make anyway.

Besides,  it is only a 2-line entry in the Makefile. So is not a lot of code to maintain.
```
# Build the docker image
docker-build: test
	docker build . -t ${IMG}
```

> 
> FWIW, I'm pro getting rid of the `make docker-build` command.
> 
> 1. It's more typing than the native docker command.
> 2. If you do know docker, you already know to run `docker build -t gcr.io/forgerock-io/secret-agent:latest .`
> 3. If you don't know enough about docker, learning our make command won't help you learn one of the most basic things about using docker.
> 4. We don't have to support that part of the makefile.
> 5. Giving people two ways to do the same thing can increase confusion or just take extra energy wondering if they're different somehow.

**Comment by snarlysodboxer** [comment]
> I would argue one is more likely to mess up (forgetting to generate the CRDs or something along those lines) by not using automation. Yes you can run `make test` + `docker build .... ` to achieve the same result. However, that ends up being more steps and you're still using `make` anyway.

Yeah, I guess it's a workflow thing. One way to do automation is to setup your editor to continuously run `go fmt` and `go vet`  as you save and type, and practice TDD and run the tests before and after every change. (`space + t` runs the tests right in vim for me.) Also, if you modify the `_types.go` file but forget to run `controller-gen` you will get compile errors. You could still forget to update your manifests but then your new feature wouldn't work so you'd quickly discover the error, and maybe learn something, instead of it happening magically behind the scenes.

**Pull request merged**